dev-tool/src/list_net/data_source.rs

```diff
@@ -0,0 +1,567 @@
+    SE: StdError + 'static,
+{
+    /// Unusable evaluation split: Assertion `split >= 0 && split.is_normal()` failed (split = {0}).
+    BadEvaluationSplit(f32),
+    /// Unusable evaluation split: With evaluation split {0} no samples are left for training.
+    ToLargeEvaluationSplit(f32),
+    /// Unusable evaluation split: Assertion `nr_evaluation_samples > 0 || split == 0` failed (split = {0}).
+    NoEvaluationSamples(f32),
+    /// A batch size of 0 is not usable for training.
+    BatchSize0,
+    /// The batch size is larger then the number of samples.
+    ToLargeBatchSize(usize),
+    /// Not reset/initialized.
+    ResetWasNotCalledBeforeTraining,
+    /// Empty database can not be used for training.
+    EmptyDatabase,
+    /// Fetching sample from storage failed: {0}.
+    Storage(SE),
+}
+
+impl<S> list_net::DataSource for DataSource<S>
+where
+    S: Storage,
+{
+    type Error = DataSourceError<S::Error>;
+
+    fn reset(&mut self, batch_size: usize) -> Result<usize, Self::Error> {
+        if batch_size == 0 {
+            return Err(DataSourceError::BatchSize0);
+        }
+        let mut rng = rand::thread_rng();
+        self.batch_size = Some(batch_size);
+        self.training_data_order.reset(&mut rng);
+        self.evaluation_data_order.reset(&mut rng);
+
+        let nr_batches = self.training_data_order.number_of_batches(batch_size);
+        if nr_batches == 0 {
+            return Err(DataSourceError::ToLargeBatchSize(nr_batches));
+        }
+        Ok(nr_batches)
+    }
+
+    fn next_training_batch(&mut self) -> Result<Vec<Sample>, Self::Error> {
+        let batch_size = self
+            .batch_size
+            .ok_or(DataSourceError::ResetWasNotCalledBeforeTraining)?;
+
+        let ids = self.training_data_order.next_batch(batch_size);
+        if ids.is_empty() {
+            return Ok(Vec::new());
+        }
+        self.storage
+            .load_batch(&ids)
+            .map_err(DataSourceError::Storage)
+    }
+
+    fn next_evaluation_sample(&mut self) -> Result<Option<Sample>, Self::Error> {
+        if let Some(id) = self.evaluation_data_order.next() {
+            match self.storage.load_sample(id) {
+                Ok(sample) => Ok(Some(sample)),
+                Err(error) => Err(DataSourceError::Storage(error)),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+pub(crate) trait Storage {
+    type Error: StdError + 'static;
+
+    /// Return the all sample ids.
+    ///
+    /// For now this is a range. I.e. `0..nr_of_samples`. We might need
+    /// to change this in the future, but for now this is simpler.
+    fn data_ids(&self) -> Result<RangeTo<usize>, Self::Error>;
+
+    /// Loads a sample and returns a reference to it.
+    ///
+    /// This method is `&mut self` as it might change the internal state
+    /// (due to loading/caching) and we also want to make sure that there
+    /// are no more lend out samples, when `load_sample` or `load_batch` are
```

Comment:
`load_sample` and `load_batch` return references to the underlying data.

When I wrote this trait I expected that there is a reasonable chance that I might have to add more complex implementations of `Storage` then a in-memory storage in the future. In which case you like would have lend
out references into a cache of some form, which will potentially flush out old samples when `load_samples`/`load_batch` are called. (Hence why they are called `load_*` instead of e.g. `get_*`.)

With the knowledge I have now, multiple weeks after writing this, I'm pretty sure that it's not needed and I'm considering to remove the `Storage` trait and just fixate `DataSource` to always use in-memory storage. (In a follow up PR, not in this PR).

Reply:
re the follow-up comment, sounds good.

Reply:
I already did it in a follow up PR by now.